### PR TITLE
[Do not merge] Migration diff upstream acs-engine master

### PR DESCRIPTION
**Pull request by yolocs**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by yolocs** [comment]
@robbiezhang @wenwu449 @yangl900

File: parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 1]
```diff
@@ -43,6 +43,11 @@ spec:
     matchLabels:
       k8s-app: kube-dns
       version: v20
+  strategy:
+    rollingUpdate:
+      maxSurge: 1
+      maxUnavailable: 1
+    type: RollingUpdate
   template:
```
Comment (wenwu449) [inline parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml:51]: acc only

File: parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 2]
```diff
@@ -168,3 +173,4 @@ spec:
       serviceAccountName: kube-dns
       nodeSelector:
         beta.kubernetes.io/os: linux
+        agentpool: system
```
Comment (wenwu449) [inline parts/k8s/addons/kubernetesmasteraddons-kube-dns-deployment.yaml:176]: acc only

File: parts/k8s/addons/kubernetesmasteraddons-kube-proxy-daemonset.yaml [hunk 3]
```diff
@@ -28,9 +28,6 @@ spec:
         - "--feature-gates=ExperimentalCriticalPodAnnotation=true"
         image: "<kubernetesHyperkubeSpec>"
         name: kube-proxy
-        resources:
-          requests:
-            cpu: 100m
         securityContext:
```
Comment (wenwu449) [inline parts/k8s/addons/kubernetesmasteraddons-kube-proxy-daemonset.yaml:31]: acc only

File: parts/k8s/kubernetesagentresourcesvmas.t [hunk 4]
```diff
@@ -27,7 +27,15 @@
...
               "primary": true,
               {{end}}
+              {{if eq $.Name "system"}}
+              "privateIPAddress": "[concat(variables('masterFirstAddrPrefix'), copyIndex(add(50, int(variables('masterFirstAddrOctet4')))))]",
+              "privateIPAllocationMethod": "Static",
+              {{else if eq $.Name "agentpool1"}}
+              "privateIPAddress": "[concat(variables('masterFirstAddrPrefix'), copyIndex(add(100, int(variables('masterFirstAddrOctet4')))))]",
+              "privateIPAllocationMethod": "Static",
+              {{else}}
               "privateIPAllocationMethod": "Dynamic",
```
Comment (wenwu449) [inline parts/k8s/kubernetesagentresourcesvmas.t:37]: acc only

File: parts/k8s/kubernetesmastercustomscript.sh [hunk 5]
```diff
@@ -56,11 +56,7 @@ echo `date`,`hostname`, startscript>>/opt/m
 # A delay to start the kubernetes processes is necessary
 # if a reboot is required.  Otherwise, the agents will encounter issue:
 # https://github.com/kubernetes/kubernetes/issues/41185
-if [ -f /var/run/reboot-required ]; then
-    REBOOTREQUIRED=true
-else
-    REBOOTREQUIRED=false
-fi
+REBOOTREQUIRED=false
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastercustomscript.sh:59]: need to make it configurable via apimodel

File: parts/k8s/kubernetesmastercustomscript.sh [hunk 6]
```diff
@@ -742,6 +744,8 @@ if [[ ! -z "${MASTER_NODE}" ]]; then
     ensureEtcd
     ensureK8s
     ensurePodSecurityPolicy
+
+    /usr/local/bin/kubectl create clusterrolebinding superuser --clusterrole=cluster-admin --user=client
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastercustomscript.sh]: acc only? still needed? @robbiezhang

File: parts/k8s/kubernetesmasterresources.t [hunk 7]
```diff
@@ -5,8 +5,8 @@
       "name": "[variables('masterAvailabilitySet')]",
       "properties":
         {
-            "platformFaultDomainCount": "2",
-            "platformUpdateDomainCount": "3",
+            "platformFaultDomainCount": "1",
+            "platformUpdateDomainCount": "1",
 		        "managed" : "true"
```
Comment (wenwu449) [inline parts/k8s/kubernetesmasterresources.t:10]: make configurable

File: parts/k8s/kubernetesmasterresources.t [hunk 8]
```diff
@@ -79,36 +94,6 @@
...
-              "destinationAddressPrefix": "*",
-              "destinationPortRange": "22-22",
-              "direction": "Inbound",
-              "priority": 101,
-              "protocol": "Tcp",
-              "sourceAddressPrefix": "*",
-              "sourcePortRange": "*"
-            }
-          },
           {
```
Comment (wenwu449) [inline parts/k8s/kubernetesmasterresources.t:97]: make configurable

File: parts/k8s/kubernetesmastervars.t [hunk 9]
```diff
@@ -211,10 +211,10 @@
...
+    "provisionScriptParametersCommon": "[concat('KUBELET_PRIVATE_KEY=',variables('clientPrivateKey'),' NETWORK_POLICY=',variables('networkPolicy'),' APISERVER_PUBLIC_KEY=',variables('apiserverCertificate'),' MAX_PODS=',variables('maxPods'),' CONTAINER_RUNTIME=',variables('containerRuntime'))]",
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastervars.t:214]: configurable

File: parts/k8s/kubernetesmastervars.t [hunk 10]
```diff
@@ -269,7 +269,7 @@
     "nsgName": "[concat(variables('agentNamePrefix'), 'nsg')]",
 {{end}}
     "nsgID": "[resourceId('Microsoft.Network/networkSecurityGroups',variables('nsgName'))]",
-    "primaryAvailabilitySetName": "[concat('{{ (index .AgentPoolProfiles 0).Name }}-availabilitySet-',variables('nameSuffix'))]",
+    "primaryAvailabilitySetName": "[concat('{{ (index .AgentPoolProfiles 1).Name }}-availabilitySet-',variables('nameSuffix'))]",
```
Comment (wenwu449) [inline parts/k8s/kubernetesmastervars.t:272]: acc only

File: parts/k8s/kuberneteswinagentresourcesvmas.t [hunk 11]
```diff
@@ -249,7 +282,7 @@
...
+          "commandToExecute": "[concat('powershell.exe -ExecutionPolicy Unrestricted -command \"', '$arguments = ', variables('singleQuote'),'-MasterIP ',variables('kubernetesAPIServerIP'),' -KubeDnsServiceIp ',variables('kubeDnsServiceIp'),' -MasterFQDNPrefix ',variables('masterFqdnPrefix'),' -Location ',variables('location'),' -AgentKey ',variables('clientPrivateKey'),' -AzureHostname ',variables('{{.Name}}VMNamePrefix'),copyIndex(variables('{{.Name}}Offset')),variables('singleQuote'), ' ; ', variables('windowsCustomScriptSuffix'), '\" > %SYSTEMDRIVE%\\AzureData\\CustomDataSetupScript.log 2>&1')]"
```
Comment (wenwu449) [inline parts/k8s/kuberneteswinagentresourcesvmas.t:285]: configurable

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 12]
```diff
@@ -29,15 +29,7 @@ param(
...
-    $AzureHostname,
-
-    [parameter(Mandatory=$true)]
-    [ValidateNotNullOrEmpty()]
-    $AADClientId,
-
-    [parameter(Mandatory=$true)]
-    [ValidateNotNullOrEmpty()]
-    $AADClientSecret
+    $AzureHostname
```
Comment (wenwu449) [inline parts/k8s/kuberneteswindowssetup.ps1:32]: configurable

File: pkg/acsengine/addons.go [hunk 13]
```diff
@@ -16,11 +16,6 @@ type kubernetesFeatureSetting struct {
 
 func kubernetesAddonSettingsInit(profile *api.Properties) []kubernetesFeatureSetting {
 	return []kubernetesFeatureSetting{
-		{
-			"kubernetesmasteraddons-heapster-deployment.yaml",
-			"kube-heapster-deployment.yaml",
-			true,
-		},
 		{
```
Comment (wenwu449) [inline pkg/acsengine/addons.go]: #18 changed this part, still need to be configurable

**Review by wenwu449 — COMMENTED** [review]
discussion to decide scope

File: parts/k8s/addons/kubernetesmasteraddons-kube-proxy-daemonset.yaml (hunk 3 above)
Comment (robbiezhang) [reply parts/k8s/addons/kubernetesmasteraddons-kube-proxy-daemonset.yaml:31]: side question: shall we add agentpool: agentpool1 to the nodeSelector?

**Comment by wenwu449** [comment]
Close this as PR #17 is merged.

**Pull request closed without merging**